server/src/main/java/org/elasticsearch/search/SearchHits.java

```diff
@@ -63,39 +68,100 @@ public SearchHits(
         @Nullable SortField[] sortFields,
         @Nullable String collapseField,
         @Nullable Object[] collapseValues
+    ) {
+        this(
+            hits,
+            totalHits,
+            maxScore,
+            sortFields,
+            collapseField,
+            collapseValues,
+            hits.length == 0 ? ALWAYS_REFERENCED : LeakTracker.wrap(new AbstractRefCounted() {
+                @Override
+                protected void closeInternal() {
+                    for (int i = 0; i < hits.length; i++) {
+                        assert hits[i] != null;
+                        hits[i].decRef();
+                        hits[i] = null;
+                    }
+                }
+            })
```

Comment:
The problem is that we add loads more ref-counting and code if we incref every hit we pass to the constructor of `SearchHits` and dec-ref it where it's constructed. So I went with the pattern of taking ownership here, just like I did for the m-search response for example. I think going forward it would actually make sense to try and move more code to this pattern as an optimisation. Maybe we could even make more stuff closeable instead of ref-counted that way and find more optimisations that way :)